pyscriptjs/src/main.ts

```diff
@@ -61,46 +92,117 @@ class PyScriptApp {
         logger.info('config loaded:\n' + JSON.stringify(this.config, null, 2));
     }
 
-    initialize() {
-        addInitializer(this.loadPackages);
-        addInitializer(this.loadPaths);
-        this.loadRuntimes();
+    // lifecycle (3)
+    showLoader() {
+        // add loader to the page body
+        logger.info('add py-loader');
+        customElements.define('py-loader', PyLoader);
+        this.loader = <PyLoader>document.createElement('py-loader');
+        document.body.append(this.loader);
     }
 
-    loadPackages = async () => {
+    // lifecycle (4)
+    loadRuntime() {
+        logger.info('Initializing runtime');
+        if (this.config.runtimes.length == 0) {
+            showError("Fatal error: config.runtimes is empty");
+            return;
+        }
+
+        if (this.config.runtimes.length > 1) {
+            showError("Multiple runtimes are not supported yet. " +
+                      "Only the first will be used");
+        }
+        const runtime_cfg = this.config.runtimes[0];
+        const runtime: Runtime = new PyodideRuntime(this.config, runtime_cfg.src,
+                                                    runtime_cfg.name, runtime_cfg.lang);
+        this.loader.log(`Downloading ${runtime_cfg.name}...`);
+        const script = document.createElement('script'); // create a script DOM node
+        script.src = runtime.src;
+        script.addEventListener('load', () => {
+            void this.afterRuntimeLoad(runtime);
+        });
+        document.head.appendChild(script);
+    }
+
+    // lifecycle (5)
+    // See the overview comment above for an explanation of how we jump from
+    // point (4) to point (5).
+    //
+    // Invariant: this.config and this.loader are set and available.
+    async afterRuntimeLoad(runtime: Runtime): Promise<void> {
+        // XXX what is the JS/TS standard way of doing asserts?
```

Comment:
✅  https://caniuse.com/mdn-api_console_assert 97%

Reply:
yes, basically my question was whether this is the idiomatic way to do it or there are better ways.
For example, I noticed that `node` has also a global `assert` function which works differently, i.e. it raises an exception instead of just writing to the console (which seems more reasonable to me), but it doesn't work in the browser.
I suppose I could write my own assert function, but wondered whether there is any standard/idiomatic/library way to do it better

Reply:
ES2+ asserts :)
```js
if( this.config !== undefined ){ throw new Error('Fatal: no config')}
if( this.loader !== undefined ){ throw new Error('Fatal: no loader')}
```